lib/store.js

```diff
@@ -0,0 +1,141 @@
+/*!
+ * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+const async = require('async');
+const bedrock = require('bedrock');
+const database = require('bedrock-mongodb');
+const jwt = require('jsonwebtoken');
+const BedrockError = bedrock.util.BedrockError;
+const NamespaceRsa = require('./namespace-rsa');
+const NamespaceHmac = require('./namespace-hmac');
+
+const logger = bedrock.loggers.get('app');
+
+class Store {
+  constructor() {
+    this.namespaces = {};
+  }
+
+  getState(namespace, callback) {
+    database.collections.jwtKeyStore.findOne(
+      {id: database.hash(namespace)}, {state: true}, (err, record) => {
+        if(err) {
+          logger.error('Mongo error when trying to find key state.', err);
+          return callback(new BedrockError(
+            'Failed to find key state due to internal error.',
+            'InternalError',
+            {namespace: namespace, httpStatusCode: 500, 'public': true}, err));
+        }
+        if(!record) {
+          return callback(new BedrockError(
+            'Key state not found.',
+            'NotFound',
+            {namespace: namespace, httpStatusCode: 404, 'public': true}));
+        }
+        callback(null, record.state);
+      });
+  }
+
+  /**
+   * Initialize a namespaced keystore.
+   *
+   * @param namespace the namespace to initialize.
+   * @param options the options to use:
+   *   sign the signing options:
+   *     algorithm the JWT signing algorithm.
+   *     tokenTtlInSecs the JWT token TTL.
+   *   [key] the key options:
+   *     id the key ID.
+   *   [verify] the verify options:
+   *     algorithms an array of acceptable HMAC signature algorithms.
+   *     clockToleranceInSecs the acceptable clock skew.
+   * @param callback(err) called once the operation completes.
+   */
+  provision(namespace, options, callback) {
+    let ns;
+    if(namespace === 'RSA') {
+      this.namespaces[namespace] = ns = new NamespaceRsa();
+    } else if(namespace === 'HMAC') {
+      this.namespaces[namespace] = ns = new NamespaceHmac();
+    } else {
+      return callback(new BedrockError(
+        'Invalid namespace specified.', 'InvalidNamespace',
+        {public: false, namespace: namespace}));
+    }
+    async.auto({
+      createState: callback => ns.createState(options, callback),
+      storeState: ['createState', (callback, results) => {
+        const id = results.createState.id;
+        const now = Date.now();
+        const record = {
+          id: database.hash(id),
+          meta: {
+            created: now,
+            updated: now
+          },
+          state: results.createState
+        };
+        database.collections.jwtKeyStore.insert(
+          record, database.writeOptions, (err, result) => {
+            if(err) {
+              if(database.isDuplicateError(err)) {
+                // namespace is already provisioned
+                return callback();
```

Comment:
Need a `TODO: ensure record in database matches given options`. Which may mean that `.insert` should actually be called from the child namespace classes, but we can deal with that later when we deal with the TODO.